core/pkg/artifacts/file_cache.go

```diff
@@ -0,0 +1,147 @@
+package artifacts
+
+import (
+	"crypto/md5"
+	"encoding/base64"
+	"io"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	atomicfile "github.com/natefinch/atomic"
+	"github.com/wandb/wandb/core/pkg/utils"
+)
+
+type FileCache struct {
+	root    string
+	enabled bool
+}
+
+func NewFileCache(enabled bool) FileCache {
+	if !enabled {
+		return FileCache{enabled: false}
+	}
+	dir, found := os.LookupEnv("WANDB_CACHE_DIR")
+	if !found {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			slog.Error("Unable to find cache directory", "err", err)
+			return FileCache{enabled: false}
+		}
+		dir = filepath.Join(userCacheDir, "wandb")
+	}
+	return FileCache{root: filepath.Join(dir, "artifacts"), enabled: true}
+}
+
+// AddFile copies a file into the cache and returns the B64MD5 cache key.
+func (c *FileCache) AddFile(path string) (string, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	return c.Write(f)
+}
+
+// AddFileAndCheckDigest copies a file into the cache. If a digest is provided, it also
+// verifies that the file's MD5 hash matches the digest.
+//
+// This is a convenience function intended for running in the background; instead of
+// returning an error it just logs it. This is often correct, since failing to add a
+// file to the cache shouldn't be a critical error.
+func (c *FileCache) AddFileAndCheckDigest(path string, digest string) {
+	if !c.enabled {
+		return
+	}
+	b64md5, err := c.AddFile(path)
+	if err != nil {
+		slog.Error("Unable to add file to cache", "err", err)
+	}
+	if digest != "" && digest != b64md5 {
+		slog.Error("File cache mismatch", "path", path, "expected", digest, "actual", b64md5)
+	}
+}
+
+// RestoreTo tries to restore the file referenced in a manifest entry to the given destination.
+//
+// If the file exists, it will be hashed and overwritten if the hash is different; if the
+// hash if correct, RestoreTo leaves it alone and returns true even if the cache is disabled.
+func (c *FileCache) RestoreTo(entry ManifestEntry, dst string) bool {
+	b64md5, err := utils.ComputeFileB64MD5(dst)
+	if err == nil && b64md5 == entry.Digest {
+		return true
+	}
+	if !c.enabled {
+		return false
+	}
+	cachePath, err := c.md5Path(entry.Digest)
+	if err != nil {
+		return false
+	}
+	exists, err := utils.FileExists(cachePath)
```

Comment:
is this check vaulable? will CopyFile return the right error if it doesn't exist? Since the cache is not atomic, between this check and line 86, we can fail regardless so might as well skip it?

Reply:
Good catch, I agree